libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,200 @@
+	}
+
+	// Add all InternalCreationDispatch calls to the RuntimeDispatch as well
+	builder.add(SpecialNode::InternalDispatch, SpecialNode::InternalCreationDispatch);
+
+	if (_contract.fallbackFunction())
+		builder.add(SpecialNode::Entry, _contract.fallbackFunction());
+
+	if (_contract.receiveFunction())
+		builder.add(SpecialNode::Entry, _contract.receiveFunction());
+
+	return std::move(builder.m_graph);
+}
+
+bool FunctionCallGraphBuilder::visit(Identifier const& _identifier)
+{
+	if (auto const* callable = dynamic_cast<CallableDeclaration const*>(_identifier.annotation().referencedDeclaration))
+	{
+		solAssert(*_identifier.annotation().requiredLookup == VirtualLookup::Virtual, "");
+
+		auto funType = dynamic_cast<FunctionType const*>(_identifier.annotation().type);
+
+		// For events kind() == Event, so we have an extra check here
+		if (funType && funType->kind() == FunctionType::Kind::Internal)
+		{
+			processFunction(callable->resolveVirtual(*m_contract), _identifier.annotation().calledDirectly);
+
+			solAssert(m_currentNode.has_value(), "");
+		}
+	}
+
+	return true;
+}
+
+bool FunctionCallGraphBuilder::visit(NewExpression const& _newExpression)
+{
+	if (ContractType const* contractType = dynamic_cast<ContractType const*>(_newExpression.typeName().annotation().type))
+		m_graph->createdContracts.emplace(&contractType->contractDefinition());
+
+	return true;
+}
+
+void FunctionCallGraphBuilder::endVisit(MemberAccess const& _memberAccess)
+{
+	auto functionType = dynamic_cast<FunctionType const*>(_memberAccess.annotation().type);
+	auto functionDef = dynamic_cast<FunctionDefinition const*>(_memberAccess.annotation().referencedDeclaration);
+	if (!functionType || !functionDef || functionType->kind() != FunctionType::Kind::Internal)
+		return;
+
+	// Super functions
+	if (*_memberAccess.annotation().requiredLookup == VirtualLookup::Super)
+	{
+		if (auto const* typeType = dynamic_cast<TypeType const*>(_memberAccess.expression().annotation().type))
+			if (auto const contractType = dynamic_cast<ContractType const*>(typeType->actualType()))
+			{
+				solAssert(contractType->isSuper(), "");
+				functionDef =
+					&functionDef->resolveVirtual(
+						*m_contract,
+						contractType->contractDefinition().superContract(*m_contract)
+					);
+			}
+	}
+	else
+		solAssert(*_memberAccess.annotation().requiredLookup == VirtualLookup::Static, "");
+
+	processFunction(*functionDef, _memberAccess.annotation().calledDirectly);
+	return;
+}
+
+void FunctionCallGraphBuilder::endVisit(ModifierInvocation const& _modifierInvocation)
+{
+	if (auto const* modifier = dynamic_cast<ModifierDefinition const*>(_modifierInvocation.name().annotation().referencedDeclaration))
+	{
+		if (*_modifierInvocation.name().annotation().requiredLookup == VirtualLookup::Virtual)
+			modifier = &modifier->resolveVirtual(*m_contract);
+		else
+			solAssert(*_modifierInvocation.name().annotation().requiredLookup == VirtualLookup::Static, "");
+
+		processFunction(*modifier);
```

Comment:
Doesn't this need to go through some kind of dispatch in the virtual lookup case?

Reply:
Hmm in this case I can't use `calledDirectly` because only Expressions have that annotation member. I wonder if we should add it there.
For this particular case I suppose I could just use `requiredLookup != Virtual` as parameter for the `processFunction` function

Reply:
I'd need to look into it, but I think you can override modifiers using inheritance which pretty much means we need some indirection here... ``requiredLookup != Virtual`` in place of ``calledDirectly`` sounds about right... but I'd need to have another look and think through this to be sure :-).

Reply:
This probably does not need to go through any dispatch by the way.

Reply:
Oh, so this is where setting `_calledDirectly` based on `requiredLookup` came from :)

So, modifiers (even virtual ones) never go through internal dispatch. Virtual function calls don't go through dispatch either - they're are resolved to a specific function before being added to dispatch. I worked on that part in the IR generator so I remember it from there. I have also added tests for this now and it only started to work like I expect it to after I changed this parameter to `true`.